Toggle bug in ProductCatalogService for demos

MrTrustor opened the pull request:
Fixes #84 

@ahmetb let me know what you think. I've used 150 microseconds because 200 slows the website down too much.

ahmetb commented on src/productcatalogservice/server.go:
```diff
@@ -148,6 +168,22 @@ type productCatalog struct{}
 	var cat pb.ListProductsResponse
 
+	if problem {
+		done := make(chan int)
```
why not just `time.Sleep()` here? Are you trying to trigger a 100% memory?

If so, please simplify as 

	t := time.Now().Add(150 * time.Millisecond)
	for time.Now().Before(t) {}

MrTrustor replied:
Yes, that's what I'm trying to do. Simplified as suggested.

ahmetb commented on src/productcatalogservice/server.go:
```diff
@@ -45,6 +47,8 @@ var (
 
 	port = flag.Int("port", 3550, "port to listen at")
+
+	problem bool = false
```
just `problem bool` is ok

MrTrustor commented:
I switched back to using USR2 to remove the bug. I don't have a strong opinion on this, but if you think that `kubectl exec` is not reliable (personally never had an issue with it), it's probably better like this.

ahmetb commented on src/productcatalogservice/README.md:
```diff
@@ -3,3 +3,23 @@
+# Trigger bug
+kubectl exec \
+    $(kubectl get pods -l app=productcatalogservice -o jsonpath='{.items[0].metadata.name}') \
+    -c server -- kill -USR1 1
```
We're about to add a entrypoint wrapper for Stackdriver Debugger to all these processes in #87 . I am not sure the debugger process is able to forward signals to its child process adequately. Do you mind testing?

MrTrustor replied:
Will do and report back.

ahmetb replied:
Thanks, that'll be great. I am sorry this is very disrupting.

MrTrustor replied:
So, I'm having a hard time figuring out what's happening here:
- The debugger seems to slow down significatively productcatalogservice in any case.
- It also seems to break the logging: I'm getting the first few lines of logs of productcatalogservice, but then nothing, even if I send directly a USR1 or USR2 to the productcatalogservice PID.
- I looked a little bit at the [code of the debugger](https://github.com/GoogleCloudPlatform/google-cloud-go/tree/master/cmd/go-cloud-debug-agent), and it doesn't look like it's forwarding USR1 and USR2. 

We can work around the fact that the debugger doesn't forward the signals, but the slowness introduced and the fact that we don't have logs anymore is a problem.

I'll open a couple of issues in [google-cloud-go](https://github.com/GoogleCloudPlatform/google-cloud-go).

MrTrustor commented on src/productcatalogservice/server.go:
```diff
@@ -148,6 +168,12 @@ type productCatalog struct{}
+	if problem {
+		t := time.Now().Add(150 * time.Microsecond)
+		for time.Now().Before(t) {
+		}
```
@ahmetb Wait before merging this, I'm getting a feedback that this version of the bug changes how Stackdriver Profiler shows it. I need to retest that.

ahmetb replied:
ACK. I don't know what to do about the debugger issues yet. Let's hold until we figure out a way forward with or without the debugger.

MrTrustor replied:
Ok, so I now better understand what's happening:

- That version of the bug doesn't really slow down the application.
- The previous version (with the `select`) actually slows down significantly the application, even with the same delay, but for a different reason: it consumes a lot more CPU, and with the load-generator running, we're running out of resources.

What do you think? I'd say the simplest is just to go back to that version:
```
	if problem {
		done := make(chan int)
		go func() {
			for {
				select {
				case <-done:
					return
				default:
				}
			}
		}()
		time.Sleep(50 * time.Microsecond)
		close(done)
	}
```

ahmetb replied:
Sounds good. But I still don’t quite understand why we are spinning another goroutine rather than blocking the current one. Feel free to restore the old code.

ahmetb commented:
We decided not to instrument Go services with debugger just yet because of the issues you pointed out. If the code above is good to you, I'll take this in.

MrTrustor commented:
@ahmetb I finally took the time to come back to this. I changed my thinking on it.
The delay is now implemented as a broken "reload catalog" feature that actually reads and parse the full catalog on each request.
It introduces a noticeable delay in the frontend and does show up in Stackdriver Profiler.

It still is triggered and removed with USR1/USR2.

Let me know what you think.

ahmetb commented on src/productcatalogservice/server.go:
```diff
@@ -147,13 +171,22 @@ func initProfiling(service, version string) {
+			log.Fatalf("failed to open product catalog json file: %v", err)
+			return nil
+		}
+		if err := jsonpb.Unmarshal(bytes.NewReader(c), &catalog); err != nil {
```
this is duplicating the code above, including the file name. so it results in less maintainable code

can we have a method named `readCatalogFile()` referenced from both places instead?

ahmetb commented:
If that gets you what you want, sounds good to me. Left a code comment.

ahmetb commented:
Looks good. Thanks.
